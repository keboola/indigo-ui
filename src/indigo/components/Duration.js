import React from 'react';
import PropTypes from 'prop-types';

export const durationFrom = (started, finished) => {
  const duration = (new Date(finished).getTime() - new Date(started).getTime()) / 1000;
  return duration;
};

export const timeInWords = (duration, round) => {
  const days = Math.floor(duration / 86400);
  const hours = Math.floor((duration % 86400) / 3600);
  const minutes = Math.floor((duration % 3600) / 60);
  const seconds = duration % 60;
  if (days > 0) {
    return 'more than 24 hrs';
  }
  if (duration === 0) {
    return '0 sec';
  }
  const result = [];
  if (hours === 1) {
    result.push(`${hours} hr`);
  }
  if (hours > 1) {
    result.push(`${hours} hrs`);
  }
  if (minutes > 0) {
    result.push(`${minutes} min`);
  }
  if (seconds > 0 && (!round || hours === 0)) {
    result.push(`${seconds} sec`);
  }
  if (result.length > 0) {
    return result.join(' ');
  }
  return '';
};

class Duration extends React.Component {
  constructor(props) {
    super(props);
    this.tick = this.tick.bind(this);
    this.state = {
      endTime: new Date().toString(),
    };
  }

  componentDidMount() {
    this.interval = setInterval(this.tick, 1000);
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  tick() {
    this.setState({
      endTime: new Date().toString(),
    });
  }

  render() {
    if (!this.props.endTime) {
      return (
        <span>
          {!!this.props.showIcon && <i className="fa fa-clock-o duration-icon" />}
          {timeInWords(durationFrom(this.props.startTime, this.state.endTime), true)}
        </span>
      );
    }

    return (
      <span>
        {!!this.props.showIcon && <i className="fa fa-clock-o duration-icon" />}
        {timeInWords(durationFrom(this.props.startTime, this.props.endTime), true)}
      </span>
    );
  }
}

Duration.propTypes = {
  startTime: PropTypes.string,
  endTime: PropTypes.string,
  showIcon: PropTypes.bool,
};

Duration.defaultProps = {
  showIcon: false,
};

export default Duration;
