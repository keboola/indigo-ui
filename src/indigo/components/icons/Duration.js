// Autogenerated file. Check svg-react-transformer.config.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const Duration = ({ className, ...props }) => (
  <svg
    width="32"
    height="32"
    viewBox="0 0 9 15"
    className={classNames('icon', className)}
    {...props}
  >
    <path
      d="M0 0v4.5h.007L0 4.508 3 7.5l-3 3 .007.008H0V15h9v-4.492h-.007L9 10.5l-3-3 3-2.992-.007-.008H9V0H0zm7.5 10.875V13.5h-6v-2.625l3-3 3 3zm-3-3.75l-3-3V1.5h6v2.625l-3 3z"
      fill="currentColor"
      fillRule="evenodd"
    />
  </svg>
);

Duration.propTypes = {
  className: PropTypes.string,
};

export default Duration;
