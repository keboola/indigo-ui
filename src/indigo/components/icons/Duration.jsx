import React from 'react';

const Duration = props => (
  <svg width={32} height={32} viewBox="0 0 9 15" {...props}>
    <path
      d="M0 0v4.5h.007L0 4.508 3 7.5l-3 3 .007.008H0V15h9v-4.492h-.007L9 10.5l-3-3 3-2.992-.007-.008H9V0H0zm7.5 10.875V13.5h-6v-2.625l3-3 3 3zm-3-3.75l-3-3V1.5h6v2.625l-3 3z"
      fill="currentColor"
      fillRule="evenodd"
    />
  </svg>
);

export default Duration;
