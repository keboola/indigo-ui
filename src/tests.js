import renderer from 'react-test-renderer';

export const snapshot = (component) => {
  const tree = renderer
    .create(component)
    .toJSON();
  expect(tree).toMatchSnapshot();
};
