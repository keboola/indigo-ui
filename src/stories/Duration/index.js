import React from 'react';
import {storiesOf} from '@storybook/react';
import {withInfo} from '@storybook/addon-info';

import Duration from '../../indigo/components/Duration';

const now = new Date();
const demoDatetime1 = now.toString();
const demoDatetime2 = new Date(now.getTime() + 123456).toString();
const demoDatetime3 = new Date(now.getTime() + 123456789).toString();
const demoSecondCount = 1586.36943;

storiesOf('Duration', module)
  .add(
    'Duration',
    withInfo({
      text: `
      Static duration component. Renders duration from start till end time given. Accepts props in datetime string format.
      `,
      inline: true,
    })(() => {
      return <Duration startTime={demoDatetime1} endTime={demoDatetime2}/>;
    })
  )
  .add(
    'Duration Dynamic',
    withInfo({
      text: `
      Active duration counter. Renders duration from start time till now.
      `,
      inline: true,
    })(() => {
      return <Duration startTime={demoDatetime1}/>;
    })
  )
  .add(
    'Duration only',
    withInfo({
      text: `
      Only formats and rounds for seconds count given. Accepts props in number format - number of seconds.
      `,
      inline: true,
    })(() => {
      return <Duration duration={demoSecondCount}/>;
    })
  )
  .add(
    'Duration > 24h',
    withInfo({
      text: `
      For duration longer than 24h limits the counter.
      `,
      inline: true,
    })(() => {
      return <Duration startTime={demoDatetime1} endTime={demoDatetime3}/>;
    })
  )
  .add(
    'Duration with icon',
    withInfo({
      text: `
      Adds icon in front of text.
      `,
      inline: true,
    })(() => {
      return <Duration startTime={demoDatetime1} endTime={demoDatetime2} showIcon/>;
    })
  )
  .add(
    'Duration - empty state',
    withInfo({
      text: `
      When no duration and start time given.
      `,
      inline: true,
    })(() => {
      return <Duration />;
    })
  );
