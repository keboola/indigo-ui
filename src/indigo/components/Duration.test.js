import React from 'react';
import moment from 'moment';

import { snapshot } from '../../tests';
import Duration from "./Duration";

const DATE_NOW = '2016-01-01T02:00:00+0100';
const DATE_PAST = '2016-01-01T00:00:00+0100';
const DATE_FUTURE_1 = '2016-01-01T02:42:16+0100';
const DATE_FUTURE_2 = '2016-01-21T00:00:00+0100';

global.Date.now = jest.fn(() => new Date(DATE_NOW).valueOf());

describe('<Duration />', () => {
  it('Duration', () => {
    snapshot(<Duration startTime={DATE_NOW} endTime={DATE_FUTURE_1}/>);
  });

  it('Duration Dynamic', () => {
    snapshot(<Duration startTime={DATE_PAST}/>);
  });

  it('Duration only', () => {
    snapshot(<Duration duration={1586.36943}/>);
  });

  it('Duration > 24h', () => {
    snapshot(<Duration startTime={DATE_NOW} endTime={DATE_FUTURE_2}/>);
  });

  it('Duration with icon', () => {
    snapshot(<Duration startTime={DATE_NOW} endTime={DATE_FUTURE_1} showIcon/>);
  });
  it('Duration - empty state', () => {
    snapshot(<Duration />);
  });
});
