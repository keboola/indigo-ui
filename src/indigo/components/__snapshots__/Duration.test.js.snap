// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Duration /> Duration > 24h 1`] = `
<span>
  more than 24 hrs
</span>
`;

exports[`<Duration /> Duration - empty state 1`] = `
<span>
  N/A
</span>
`;

exports[`<Duration /> Duration 1`] = `
<span>
  4 sec
</span>
`;

exports[`<Duration /> Duration Dynamic 1`] = `
<span>
  2 min 4 sec
</span>
`;

exports[`<Duration /> Duration only 1`] = `
<span>
  26 min 26 sec
  
</span>
`;

exports[`<Duration /> Duration with icon 1`] = `
<span>
  <i
    className="fa fa-clock-o duration-icon"
  />
  4 sec
</span>
`;
